### PR TITLE
Fix getOffsetForCaret to return correct value if contains widget span

### LINKED ISSUE #98458: RenderParagraph getOffsetForCaret does not work correct if it has WidgetSpan
## Steps to Reproduce

```dart
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final GlobalKey text = GlobalKey();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Text.rich(TextSpan(
        children: <InlineSpan>[
          TextSpan(text:'a'),
          WidgetSpan(child: Text('b')), // It works if comment out this line.
          TextSpan(text: 'c'),
        ]
      ), key:text,),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          RenderParagraph paragraph = text.currentContext!.findRenderObject()! as RenderParagraph;
          final String string = paragraph.text.toPlainText();
          final Offset offset = paragraph.getOffsetForCaret(TextPosition(offset: string.length, affinity: TextAffinity.upstream), Rect.zero);
          showDialog<void>(
            context: context,
            barrierDismissible: false,
            builder: (BuildContext context) {
              return AlertDialog(
                title: const Text('getOffsetForCaret'),
                content: SingleChildScrollView(
                  child: ListBody(
                    children: <Widget>[
                      Text('string is $string'),
                      Text('string length is ${string.length}'),
                      Text('getOffsetForCaret at location "${string.length}" is $offset'),
                    ],
                  ),
                ),
                actions: <Widget>[
                  TextButton(
                    child: const Text('Approve'),
                    onPressed: () {
                      Navigator.of(context).pop();
                    },
                  ),
                ],
              );
            },
          );
        },
        child: Icon(Icons.ice_skating),
      ),
    );
  }
}
```

Click the floating button to see the the result in the dialog

expect: the location is at the end of the Text
actual: the location is always (0, 0)

If comment out the widgetspan, things work as expected.

Something may be wrong with the textpainter.

### PR DESCRIPTION
**Pull request by chunhtai**
fixes https://github.com/flutter/flutter/issues/98458

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by chunhtai** [comment]
a friendly bump

File: packages/flutter/lib/src/widgets/widget_span.dart [hunk 1]
```diff
@@ -139,7 +139,8 @@ class WidgetSpan extends PlaceholderSpan {
 
   @override
   int? codeUnitAtVisitor(int index, Accumulator offset) {
-    return null;
+    offset.increment(1);
+    return PlaceholderSpan.placeholderCodeUnit.codeUnitAt(0);
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/widget_span.dart:143]: Are there any implications for doing this?  Like I wonder, in the future if we support copying a WidgetSpan, will the plain text representation of the WidgetSpan be `\uFFFC`?

**Review by justinmc — APPROVED** [review]
LGTM 👍 

Just a question below.  Also CC @Renzo-Olivares just as a heads up, since this is somewhat related to rich text editing.

File: packages/flutter/lib/src/widgets/widget_span.dart (hunk 1 above)
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/widget_span.dart:143]: This method is mainly used to calculate layout geometry. I would imagine the copying use the toPlainText instead, or something like toRichText. It has to return something when this method is called, returning null at index within a string seems counter-intuitive.

Comment (justinmc) [reply packages/flutter/lib/src/widgets/widget_span.dart:143]: Sounds good, it shouldn't matter at all for copying then.  The `\uFFFC` character makes sense to me here, I think that's what I was using as a placeholder when messing around with rich text editing.

File: packages/flutter/lib/src/painting/placeholder_span.dart [hunk 2]
```diff
@@ -41,6 +41,9 @@ abstract class PlaceholderSpan extends InlineSpan {
     TextStyle? style,
   }) : super(style: style);
 
+  /// The unicode character to represent a placeholder.
+  static const String placeholderCodeUnit = '\uFFFC';
```
Comment (tgucio) [inline packages/flutter/lib/src/painting/placeholder_span.dart:45]: Is there a reason no to define this as an int code unit = 0xFFFC?
This way you could call buffer.writeCharCode() and avoid doing the slightly confusing placeholderCodeUnit.codeUnitAt(0) conversion in widget_span.dart

Comment (tgucio) [reply packages/flutter/lib/src/painting/placeholder_span.dart:45]: Ok, I published #98971 as this PR has been merged already.

**Pull request merged**